src/ui/features/onboarding/Success/ReferralCodeWidget/ReferralCodeWidget.tsx

```diff
@@ -0,0 +1,181 @@
+          referralCode: rawReferralCode.trim(),
+        });
+      },
+      onSuccess,
+    });
+
+  return (
+    <>
+      <Spacer height={28} />
+      <VStack gap={16} className={styles.referralCodeForm}>
+        <UIText
+          kind="body/accent"
+          color="var(--always-white)"
+          style={{ textAlign: 'left' }}
+        >
+          Referral Code
+        </UIText>
+        <form
+          onSubmit={(event) => {
+            event.preventDefault();
+
+            const form = event.currentTarget;
+            if (!form.checkValidity()) {
+              return;
+            }
+            const formData = collectData(form, {});
+            const referralCode = formData.referralCode as string;
+
+            applyReferralCode({ referralCode });
+          }}
+        >
+          <VStack gap={4}>
+            <HStack gap={8} alignItems="center">
+              <ZStack>
+                <Input
+                  className={styles.referralCodeInput}
+                  name="referralCode"
+                  autoFocus={true}
+                  boxHeight={40}
+                  placeholder="Enter code"
+                  required={true}
+                  value={referralCode}
+                  onChange={(event) => setReferralCode(event.target.value)}
+                  disabled={applyReferralCodeMutation.isLoading}
+                />
+                {referralCode.length > 0 ? (
+                  <ClearSolidIcon
+                    style={{
+                      width: 20,
+                      height: 20,
+                      cursor: 'pointer',
+                      alignSelf: 'center',
+                      justifySelf: 'end',
+                      color: 'var(--neutral-500)',
+                      marginRight: 8,
+                    }}
+                    onClick={() => {
+                      setReferralCode('');
+                      applyReferralCodeMutation.reset();
+                    }}
+                  />
+                ) : null}
+              </ZStack>
+              <Button
+                kind="regular"
+                style={{
+                  padding: 8,
+                  height: 40,
+                  borderRadius: 8,
+                  ['--button-background' as string]: 'var(--always-white)',
+                  ['--button-background-hover' as string]: '#f0f0f2',
+                }}
+                disabled={applyReferralCodeMutation.isLoading}
+              >
+                <CheckIcon
+                  style={{
+                    color: 'var(--primary)',
+                    width: 24,
+                    height: 24,
+                  }}
+                />
+              </Button>
+            </HStack>
+            {applyReferralCodeMutation.isError ? (
+              <UIText
+                kind="caption/regular"
+                color="var(--negative-500)"
+                style={{ textAlign: 'left' }}
+              >
+                Invalid Referral Code
```

Comment:
Ideally we should handle network errors differently than "invalid code"

Reply:
Especially since under the hood we fetch a lot of requests. If any of those fail it's wrong to say "invalid code"

Reply:
`requestErrorToMessage`?